rust/noosphere-ns/src/helpers.rs

```diff
@@ -0,0 +1,86 @@
+use anyhow::Result;
+use libp2p::Multiaddr;
+use noosphere_core::{authority::generate_ed25519_key, data::Did, view::SPHERE_LIFETIME};
+use noosphere_storage::{MemoryStorage, SphereDb};
+use ucan::{builder::UcanBuilder, crypto::KeyMaterial, store::UcanJwtStore, Ucan};
+use ucan_key_support::ed25519::Ed25519KeyMaterial;
+
+use crate::{
+    dht::RecordValidator, utils::generate_capability, DhtConfig, NameSystem, NameSystemClient,
+};
+/// Data related to an owner sphere and a NameSystem running
+/// on its behalf in it's corresponding gateway.
+pub struct NsData {
+    pub ns: NameSystem,
+    pub owner_key: Ed25519KeyMaterial,
+    pub owner_id: Did,
+    pub sphere_id: Did,
+    pub delegation: Ucan,
+}
+
+pub async fn generate_name_system<V: RecordValidator + 'static>(
+    store: &mut SphereDb<MemoryStorage>,
+    bootstrap_addresses: &[Multiaddr],
+    validator: V,
+) -> Result<NsData> {
+    let owner_key = generate_ed25519_key();
+    let owner_id = Did(owner_key.get_did().await?);
+    let sphere_key = generate_ed25519_key();
+    let sphere_id = Did(sphere_key.get_did().await?);
+
+    // Delegate `sphere_key`'s publishing authority to `owner_key`
+    let delegate_capability = generate_capability(&sphere_id);
+    let delegation = UcanBuilder::default()
+        .issued_by(&sphere_key)
+        .for_audience(&owner_id)
+        .with_lifetime(SPHERE_LIFETIME)
+        .claiming_capability(&delegate_capability)
+        .build()?
+        .sign()
+        .await?;
+    let _ = store.write_token(&delegation.encode()?).await?;
```

Comment:
Kind of a bummer that `store` is being passed around these utils only for this side-stepping-Ipfs hack here, now that `store` is internal to the `validator`